text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+allowed one struct to embed the fields of another:
+
+```rust
+struct Foo {
+    /* Foo fields here */
+}
+
+struct Bar {
+    ..Foo,
+    /* add'l bar fields here */
+}
+```
+
+In addition, this notation could be used in traits:
+
+```rust
+trait Baz {
+    ..Bar // roughly equivalent to `bar: Bar`, but fields can be accessed directly 
+}
+```
+
+The original drafts defined `..` to imply "prefix" layout. This made
+field access through trait objects particularly efficient (no vtable
+access is needed). However, this was removed because it is unclear
+whether prefix layout (and efficient trait object access) is important
+enough to merit this syntax: perhaps it should be controlled through a
+`#[repr]` attribute. Moreover, prefix semantics are less composable
+than some other alternatives.
+
+If one does *not* tie `..Foo` to prefix semantics, but instead say
+that it just means that the trait is implemented by "some struct that
+itself includes a `..Foo` somewhere", this also raises some
+interesting questions.  For example, what are the semantics of having
+multiple instances of `..Foo` in the same struct? (Perhaps inherited
+transitively.) In working out examples, it seemed clear that there was
+enough complexity here to merit a distinct RFC (and further
+consideration).
+
+In any case, adding `..Foo` notation does not add any expressiveness.
+It can always be modeled by using explicit fields, at the cost of some
+ergonomics.
+
+### Other changes
+
+To use the pattern in [aturon's blog post][aturon] ergonomically, some
+other changes are needed; these changes are largely orthogonal to the
+current RFC. Many of these were
+[originally proposed in RFC 250][RFC250].
+
+- **Upcasting:** it must be possible to upcast a trait object into a
+  "super-trait" object. Currently this is not supported.
+  Without this, one cannot pass a `&Container` object to a function that
+  expects a `&GuiNode` object.
+- **Implicit trait coercion:** it must be possible to invoke an
+  inherent method defined on trait `Foo` when given an object of some
+  type `T: Foo`. Without this, one could not invoke inherent methods
+  defined on the `GuiNode` trait on a `Circle` struct. (This same
+  problem affects e.g. the methods on the `Any` trait object.)
+- **Thin traits:** For maximum efficiency, we should be able to
+  declare a trait as a "thin trait". This would impose stricter orphan
+  rules on the trait in exchange for making the size of a trait object
+  be a single word (because the vtable can be embedded into the
+  implementing types).
+- **Super fn definitions:** Specialization currently offers no
+  equivalent to the notion of a "super fn" definition from OO
+  languages. This means that if, e.g., `Circle` wanted to override the
+  `highlight` method but also call out to the prior version, it can't
+  easily do so. Super calls are not strictly needed thoug, as one can
+  always refactor the super call into a free fn.
```

Comment:
This section contains references to types such as `Container` and `Circle` which are not mentioned in the RFC anywhere else. Probably a left-over from previous revisions?